base/channels.jl

```diff
@@ -553,6 +704,135 @@ function take_unbuffered(c::Channel{T}) where T
+    # for small-ish n, we can avoid resizing the buffer as we push elements
+    if n < 2^16
+        resize!(buffer, n)
+    end
+
+    buffered = isbuffered(c)
+    # short-circuit for small n
+    if n == 0
+        return buffer
+    elseif n == 1
+        try
+            x = take!(c)
+            buffer[begin] = x
+        catch e
+            # when provided a batch size, take should not error on a closed channel
+            # so we need to return an empty buffer if we fail to get one element.
+            if e isa InvalidStateException && e.state === :closed
+                empty!(buffer)
+            else
+                rethrow()
+            end
+        end
+        return buffer
+    end
+    return buffered ? take_buffered(c, n, buffer) : take_unbuffered(c, n, buffer)
+end
+
+function take_buffered(c::Channel{T}, n, buffer::AbstractVector) where {T}
+    elements_taken = 0 # number of elements taken so far
+    idx1 = firstindex(buffer)
+    target_buffer_len = min(n, c.sz_max)
+    lock(c)
+    try
+        while elements_taken < n && (isopen(c) || isready(c))
+            # wait until the channel has at least min_n elements or is full
+            while length(c.data) < target_buffer_len && isopen(c)
+                try
+                    wait(c.cond_take)
+                catch e
+                    if e isa InvalidStateException && e.state === :closed
+                        break
+                    else
+                        rethrow()
+                    end
+                end
+            end
+            # take as many elements as possible from the buffer
+            n_to_take = min(n - elements_taken, length(c.data))
+            idx_start = idx1 + elements_taken
+            idx_end = idx_start + n_to_take - 1
+            if idx_end > lastindex(buffer)
+                resize!(buffer, idx_end)
+            end
+            # since idx_start/end are both created relative to `firstindex(buffer)`, they are safe to use
+            # them as indices for the buffer
+            @views copy!(buffer[idx_start:idx_end], c.data[1:n_to_take])
+            deleteat!(c.data, 1:n_to_take)
+            _increment_n_avail(c, -n_to_take)
+            foreach(_ -> notify(c.cond_put, nothing, true, false), 1:n_to_take)
+            elements_taken += n_to_take
+        end
+
+        # if we broke early, we need to remove the extra slots from the result
+        if elements_taken < length(buffer)
+            deleteat!(buffer, firstindex(buffer)+elements_taken:lastindex(buffer))
+        end
+        return buffer
+    finally
+        unlock(c)
+    end
+end
+
+function take_unbuffered(c::Channel, n, buffer::AbstractVector)
+    idx1 = firstindex(buffer)
+    last_idx = idx1 - 1
+    lock(c)
+    try
+        for i in idx1:(idx1+n-1)
+            try
+                if i <= lastindex(buffer)
+                    buffer[i] = take_unbuffered(c)
+                else
+                    push!(buffer, take_unbuffered(c))
```

Comment:
Maybe we should have some heuristic here about resizing the buffer if we are still far away from `n`